yaml.go

```diff
@@ -300,24 +301,57 @@ type yamlError struct {
 	err error
 }
 
+func (e *yamlError) Unwrap() error {
+	return e.err
+}
+
 func fail(err error) {
-	panic(yamlError{err})
+	panic(&yamlError{err})
 }
 
 func failf(format string, args ...interface{}) {
-	panic(yamlError{fmt.Errorf("yaml: "+format, args...)})
+	panic(&yamlError{fmt.Errorf("yaml: "+format, args...)})
+}
+
+// ParserError is each error with a source line position found by the parser.
+// Unlike UnmarshalErrors, it is only used for terminal failures.
+type ParserError struct {
+	Message string
+	Line    int
+}
+
+func (e *ParserError) Error() string {
+	var b strings.Builder
+	b.WriteString("yaml: ")
+	if e.Line != 0 {
+		b.WriteString("line " + strconv.Itoa(e.Line) + ": ")
+	}
+	b.WriteString(e.Message)
+	return b.String()
+}
+
+// UnmarshalError is each error with a source position found by Unmarshal.
+type UnmarshalError struct {
+	Message string
+	Line    int
+	Column  int
 }
```

Comment:
I suggest adding a way to differentiate what type of error actually happened, because not all errors have the same severity. For instance, an error because the file contains an _unknown_ field may only constitute printing a warning, while not being able to unmarshal a _known_ field would probably be reason enough to stop processing the file altogether.

I added a snippet of one possible solution that could plug into the current code. Another would be to create separate Go types for each error, although that would be a bigger change.

```suggestion
type UnmarshalError struct {
	Type    UnmarshalErrorType
	Message string
	Line    int
	Column  int
}

type UnmarshalErrorType int

const (
	UnknownField    UnmarshalErrorType = iota
	CannotUnmarshal UnmarshalErrorType
	// ...
)
```

Reply:
I would propose to do this in a separate PR as you'll also need to touch the places where these types are generated?

Reply:
First a disclaimer: I'm not a maintainer of this project, so you don't get the wrong idea, my comments don't have a lot of weight 🙂

The type would only have to be added to the lines where `UnmarshalError` is created, so all the lines that would need to change are already touched in this PR. That said, a separate PR sounds reasonable to me! I might create it and base it on your PR since I need the functionality in my project.

Reply:
I had something like this in mind: https://github.com/lovromazgon/yaml/commit/456dd7392ac149ed5db0d75568f4be5ac9a8a690

It changes `UnmarshalError` to an interface and creates specific types for each error. This allows you to know exactly what error happened and act accordingly.

I won't create a PR until @niemeyer responds to this PR though, to not create more noise and confusion.